language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -184,228 +126,471 @@ pub struct RootAccessPath {
         }
+        Ok(true)
     }
 
-    // Structural non-equality for Move values
-    // Implemented by hand instead of `!equals` to allow for short circuiting
-    pub fn not_equals(self, l2: Local) -> VMResult<bool> {
-        match (self, l2) {
-            (Local::Ref(mv1), Local::Ref(mv2)) | (Local::Value(mv1), Local::Value(mv2)) => {
-                mv1.not_equals(&mv2)
-            }
-            (Local::GlobalRef(gr1), Local::GlobalRef(gr2)) => {
-                gr1.read_reference().not_equals(&gr2.read_reference())
-            }
-            (Local::GlobalRef(gr), Local::Ref(mv)) => gr.read_reference().not_equals(&mv),
-            (Local::Ref(mv), Local::GlobalRef(gr)) => mv.not_equals(&gr.read_reference()),
-            (Local::Invalid, Local::Invalid) => Ok(false),
+    /// Normal code should always know what type this value has. This is made available only for
+    /// tests.
+    #[allow(non_snake_case)]
+    #[doc(hidden)]
+    fn to_struct_def_FOR_TESTING(&self) -> StructDef {
+        let fields = self.0.iter().map(MutVal::to_type_FOR_TESTING).collect();
+        StructDef::new(fields)
+    }
+}
+
+// Private API for a `Reference`. It is jut a pass through layer. All those should disappear
+// once compiled
+impl Reference {
+    // called from cost synthesis, revisit
+    pub fn new(value: Value) -> Self {
+        Reference(MutVal::new(value))
+    }
+
+    fn into_value(self) -> VMResult<Value> {
+        self.0.into_value()
+    }
+
+    fn copy_value(&self) -> VMResult<Value> {
+        self.0.copy_value()
+    }
+
+    fn size(&self) -> AbstractMemorySize<GasCarrier> {
+        words_in(*REFERENCE_SIZE)
+    }
+
+    fn borrow_field(&self, field_offset: usize) -> VMResult<Value> {
+        self.0.borrow_field(field_offset)
+    }
+
+    fn write_value(self, value: Value) {
+        self.0.write_value(value);
+    }
+
+    fn equals(&self, ref2: &Reference) -> VMResult<bool> {
+        self.0.equals(&ref2.0)
+    }
+
+    /// Normal code should always know what type this value has. This is made available only for
+    /// tests.
+    #[allow(non_snake_case)]
+    #[doc(hidden)]
+    fn to_type_FOR_TESTING(&self) -> Type {
+        self.0.to_type_FOR_TESTING()
+    }
+}
+
+/// Implementation for reference opcodes.
+///
+/// A reference in the runtime can have different shapes so any time a reference is expected
+/// a `ReferenceValue` is created by popping whatever reference is on the stack.
+/// Operations on the reference (via opcodes) are then invoked on a `ReferenceValue`.
+impl ReferenceValue {
+    /// Create a `ReferenceValue` from a `Value` popped off the stack.
+    /// Fails if the value is not a reference of some kind.
+    pub fn new(value: Value) -> VMResult<Self> {
+        match value.0 {
+            ValueImpl::Reference(reference) => Ok(ReferenceValue::Reference(reference)),
+            ValueImpl::GlobalRef(reference) => Ok(ReferenceValue::GlobalRef(reference)),
```

Comment:
I believe Rust will let you write  the shorter

```
ValueImpl::Reference(reference) | 
ValueImpl::GlobalRef(reference) => Ok(ReferenceValue::GlobalRef(reference)),
```
(and similar in a few cases below)

Reply:
those are not the same impl (as we discussed) but I'll look in the file if there are such cases (I thought I fixed them but...

Reply:
Argh, my bad. I was thinking of:

- Lines 567 + 568
- Lines 575 + 576
- Lines 583 + 584

Reply:
the problem is that the receiver is not the same type so the compiler barfs (as I think it should?)